### ISSUE TITLE
Reload oxAuth configuration without tomcat restart

### ISSUE THREAD

**Issue by yurem** [GluuFederation/oxAuth#22]
oxAuth should has reload configuration process which should check if there is marker file in conf folder like oxauth.config.reload

**Comment by nynymike** [comment]
Yura, this is urgently needed for clusters.

**Comment by yuriyz** [comment]
I can take over it.

**Comment by yurem** [comment]
We can add one more property version to oxauth-server.properties.
oxAuth periodically should load oxauth-server.properties and compare version in LDAP and in file. After that if should reload config if needed.

**Comment by yurem** [comment]
Also we can check if there is specified file in conf folder and reload config if this file exists.

**Comment by yurem** [comment]
@iromli  Which method is better for cluster?

**Comment by yurem** [comment]
@yuriyz Thank you for the help! Yes, we can split this task. I implemented service which check if there is required condition to restart service: https://github.com/GluuFederation/oxAuth/commit/4c11bb18d1947c05858a7664c967ac14fdc0efe3
I added new property "configuration-reload-trigger" to oxauth-server.properties.

Can you implement method where we load config from file and update in LDAP: ConfigurationFactory.reoadConfigurationImpl

I think we should try to reload new config file and only after that we need to store updated config in LDAP. This will save us from scenario where new config files are broken.

**Comment by yurem** [comment]
@iromli this is update oxauth config: https://github.com/GluuFederation/community-edition-setup/commit/22820af2c2b9bc54d92a74a318fc0e0acd5fa158

**Comment by yuriyz** [comment]
@yurem Sure ;). 
I think we need at least two possibilities to reload configuration:
a) by property from configuration itself - this is what you've added
b) by external file - to force reload from outside
Imagine that you have configuration-reload-trigger=false. and you would like to trigger reload. Then the only option is to change it in configuration itself. I guess it is also convenient to create "oxauth.conf.reload" file to force reload. At least it gives more flexibility.

**Comment by yuriyz** [comment]
or made it even simpler, timer will just check on oxauth-config.xml.reload file. Because I don't see advantages to keep reload information in configuration itself.

**Comment by yuriyz** [comment]
ok, configuration reload is implemented as following:
1) if reload marker exists ( <Tomcat>/conf/oxauth.config.reload) then oxauth starts to monitor configuration files (oxauth-config.xml, oxauth-static-conf.json, oxauth-web-keys.json, oxauth-errors.json). If marker file does not exist or deleted then oxauth stops monitoring process.
2) configuration monitor checks whether any of these files are changed. If changed then reload changed part and persist it to LDAP (to force propagation in cluster).

Please let me know whether it fits to your cluster solution.

**Comment by iromli** [comment]
@yuriyz and @yurem thanks.

I do have questions though:

1.   So the only things I need to do is updating oxauth-config.xml:

     ```xml
<configuration-reload-trigger>/opt/tomcat/conf/reload_oxauth_config_marker</configuration-reload-trigger>
     ```

     and creating `/opt/tomcat/conf/reload_oxauth_config_marker`?

2. Which oxAuth version affected by this changes? Does v2.2 also has this feature? Note, cluster still use oxAuth v2.2, but I can upgrade to latest version that has this reload feature.

**Comment by nynymike** [comment]
I think just version 2.3 because clusters needs to upgrade to 2.3 anyway

**Comment by yuriyz** [comment]
@iromli the only thing you need is to create /opt/tomcat/conf/oxauth.config.reload file. It will force oxauth to monitor all configuration files. Changes are made in master branch (version 3.0). We can integrate/merge it into 2.3 version.

**Comment by iromli** [comment]
@yuriyz 

> We can integrate/merge it into 2.3 version.

Noted. Cluster will be upgraded to 2.3.

**Comment by yurem** [comment]
Let's plan to issue 2.3.2 on next week
Maybe we will include other upgrades in this version too

**Comment by iromli** [comment]
Apology for bringing up this issue, but is this feature already shipped to v2.3?
